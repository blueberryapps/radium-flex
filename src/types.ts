export type BreakpointKey = 'xs' | 'ms' | 'sm' | 'md' | 'lg';
export type Breakpoints = {[K in BreakpointKey] : number};

export type MediaQueryKey = BreakpointKey | 'col';
export type MediaQueries = {[K in MediaQueryKey] : string};

export type ColumnSize = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export interface BoxProps {
  className?: string;
  flex?: boolean;
  order?: number;
  self?:
    'auto' |
    'stretch' |
    'center' |
    'baseline' |
    'flex-start' |
    'flex-end';
  grow?: number;
  shrink?: number;
  basis?: number | string;
  col?: ColumnSize;
  lg?: ColumnSize;
  md?: ColumnSize;
  sm?: ColumnSize;
  ms?: ColumnSize;
  xs?: ColumnSize;
  style?: Object;
  testId?: string;
  routerLink?: boolean;
  to: string;
};

export interface FlexProps {
  className?: string;
  debug?: boolean;
  direction?: 'row' | 'column';
  style?: Object;
  wrap?: 'wrap' | 'nowrap';
  gap?: number | boolean;
  align?:
    'stretch' |
    'center' |
    'baseline' |
    'flex-start' |
    'flex-end';
  content?:
    'stretch' |
    'center' |
    'space-around' |
    'space-between' |
    'flex-start' |
    'flex-end';
  justify?:
    'center' |
    'space-around' |
    'space-between' |
    'flex-start' |
    'flex-end';
};
