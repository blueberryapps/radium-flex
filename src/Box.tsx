import * as Radium from 'radium';
import * as React from 'react';
import { columns as columnsStyle, debugStyle, defaultGap, mediaQueries } from './styles';
import { BoxProps, FlexProps, MediaQueryKey } from './types';

export const ReactComponent = React.PureComponent || React.Component;
export const orderedBreakpoints = ['col', 'xs', 'ms', 'sm', 'md', 'lg'];
export type Breakpoints = 'col' | 'xs' | 'ms' | 'sm' | 'md' | 'lg';

export const defaultProps = {
  col: 12,
  debug: false,
  direction: 'row',
  gap: 0,
  grow: 0,
  self: 'auto',
  shrink: 1,
  wrap: 'wrap',
};

export const calculateColumnWidth = (props: BoxProps, size: Breakpoints): number => {
  const result = orderedBreakpoints.slice(0, orderedBreakpoints.indexOf(size) + 1).reduce(
    (acc, breakpoint) => (typeof props[breakpoint] === 'number' ? props[breakpoint] : acc),
    props.col
  );

  return typeof result === 'number' ? result : defaultProps.col;
};

export class Box extends ReactComponent<Partial<BoxProps & FlexProps>, void> {
  static defaultProps = defaultProps;

  render() {
    const { children } = this.props;
    const { className, style, testId } = this.props as BoxProps & FlexProps;

    return (
      <div className={className} data-test={testId} style={[this.columns(), this.styles(), style]}>
        {children}
      </div>
    );
  }

  getColumnSize(size: Breakpoints): number {
    const { col, xs, ms, sm, md, lg } = this.props;
    return calculateColumnWidth({ col, xs, ms, sm, md, lg } as BoxProps, size);
  }

  columns(): React.CSSProperties[] {
    return orderedBreakpoints
      .map((breakpoint: MediaQueryKey) => this.getColumnStyle(breakpoint, this.getColumnSize(breakpoint)));
  }

  getColumnStyle(breakpoint: MediaQueryKey, columns: number): React.CSSProperties {
    return { [`${mediaQueries[breakpoint]}`]: this.getColumnWidth(columns) };
  }

  getColumnWidth(column: number): React.CSSProperties {
    const { flex, columns } = this.props;
    const { floor } = Math;
    const colCount = columns || columnsStyle;

    if (column === 0) {
      return { width: 0, display: 'none' };
    }

    const width: number = column > 0 ? (floor(100000 / colCount * column) / 1000) : 100;

    return { width: `${width}%`, display: flex ? 'flex' : 'block' };
  }

  styles(): React.CSSProperties {
    const { debug, direction, flex, gap, grow, order, self, shrink, wrap } = this.props;

    const debugStyles: Object = debug ? debugStyle() : {};
    const gapSize = gap === true ? defaultGap : gap;
    const gapStyles = typeof gapSize === 'number' && gapSize > 0 ? { paddingLeft: `${gapSize}px`, paddingRight: `${gapSize}px` } : {};
    const flexStyles = flex ? {
      display: 'flex',
      flexDirection: direction,
      flexWrap: wrap,
    } : {};

    return {
      display: 'block',
      alignSelf: self,
      flexGrow: grow,
      flexShrink: shrink,
      order,
      boxSizing: 'border-box',
      width: '100%',
      ...flexStyles,
      ...gapStyles,
      ...debugStyles,
    };
  }
}

export default Radium(Box);
