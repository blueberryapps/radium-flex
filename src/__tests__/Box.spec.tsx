import { render } from 'enzyme';
importÂ  { StyleRoot } from 'radium';
import * as React from 'react';
import Box, { Breakpoints, calculateColumnWidth } from '../Box';
import { BoxProps } from '../types';

it('should render Box with default style', () => {
  expect(render(<StyleRoot><Box>content</Box></StyleRoot>)).toMatchSnapshot();
});

describe('calculate column width', () => {
  const testCases = [
    {props: {col: 1, xs: 1, ms: 1, sm: 1, md: 1, lg: 1}, size: 'ms', width: 1},
    {props: {col: 1}, size: 'ms', width: 1},
    {props: {col: 1}, size: 'lg', width: 1},
    {props: {col: 1, xs: 2, ms: 3, sm: 4, md: 5, lg: 6}, size: 'col', width: 1},
    {props: {col: 1, xs: 2, ms: 3, sm: 4, md: 5, lg: 6}, size: 'xs', width: 2},
    {props: {col: 1, xs: 2, ms: 3, sm: 4, md: 5, lg: 6}, size: 'ms', width: 3},
    {props: {col: 1, xs: 2, ms: 3, sm: 4, md: 5, lg: 6}, size: 'sm', width: 4},
    {props: {col: 1, xs: 2, ms: 3, sm: 4, md: 5, lg: 6}, size: 'md', width: 5},
    {props: {col: 1, xs: 2, ms: 3, sm: 4, md: 5, lg: 6}, size: 'lg', width: 6},
    {props: {xs: 2, md: 5}, size: 'col', width: 12},
    {props: {xs: 2, md: 5}, size: 'xs', width: 2},
    {props: {xs: 2, md: 5}, size: 'ms', width: 2},
    {props: {xs: 2, md: 5}, size: 'sm', width: 2},
    {props: {xs: 2, md: 5}, size: 'md', width: 5},
    {props: {xs: 2, md: 5}, size: 'lg', width: 5},
    {props: {col: 0, md: 5}, size: 'xs', width: 0},
    {props: {col: 0, md: 5}, size: 'ms', width: 0},
    {props: {col: 0, md: 5}, size: 'sm', width: 0},
    {props: {col: 0, md: 5}, size: 'md', width: 5},
    {props: {col: 0, md: 5}, size: 'lg', width: 5},
    {props: {col: 5, md: 0}, size: 'xs', width: 5},
    {props: {col: 5, md: 0}, size: 'ms', width: 5},
    {props: {col: 5, md: 0}, size: 'sm', width: 5},
    {props: {col: 5, md: 0}, size: 'md', width: 0},
    {props: {col: 5, md: 0}, size: 'lg', width: 0},
  ];

  testCases.forEach((testCase) => {
    it(`Calculates width: ${JSON.stringify(testCase)}`, () => {
      expect(calculateColumnWidth(testCase.props as BoxProps, testCase.size as Breakpoints))
        .toEqual(testCase.width);
    });
  });
});
