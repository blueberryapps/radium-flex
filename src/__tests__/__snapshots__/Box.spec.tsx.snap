// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Box with default style 1`] = `
<div
  data-radium="true"
>
  <div
    class="rmq-ce8f0f0d rmq-d31490d2 rmq-ed672573 rmq-5385cdb1 rmq-c847798b rmq-fe61d09d"
    data-radium="true"
    style="display:block;align-self:auto;flex-grow:0;flex-shrink:1;box-sizing:border-box;width:100%"
  >
    content
  </div>
</div>
`;
